### ISSUE TITLE
MapBox test report

### ISSUE THREAD

**Issue by jezikk82** [BimmerGestalt/AAIdrive#538]
Hi, I just drive 60+ km with MapBox without issue. Everything worked fine, even the performance was better than GMaps.

I dont know if it's possible but would be nice to have: 
1. Missing automatic rerouting. have to manually get back to main screen and press reroute.
2. Basic navi direction (e.g. turn right in 500m) and route summary (e.g. kilometers/time left to target)
3.  Search history of map destinations/places
4. Map rotation to direction we are driving, similar to any navi view and moving actual position dot to about 20% of the height (now is in the middle).   
5. More details on the map (e.g. building numbers)
6. Possibility to compile variant with GMap and MapBox

**Comment by fata74** [comment]
Hi, great implementation that works on ID4.Can't fill the full screen?
![PXL_20220223_172359278 MP](https://user-images.githubusercontent.com/56363659/155527276-69e580fd-d341-449e-8416-254baa224a25.jpg)

**Comment by fata74** [comment]
![Uploading PXL_20220223_170551187.jpg…]()

**Comment by fata74** [comment]
![PXL_20220223_170551187](https://user-images.githubusercontent.com/56363659/155527609-3802ced1-6753-47f7-ae49-033f62bc019a.jpg)

**Comment by jezikk82** [comment]
What scree resolution you have. (Can be read from app under car info - advanced - hmi.display-*: xxx).

**Comment by fata74** [comment]
![IMG_20220224_155404](https://user-images.githubusercontent.com/56363659/155548573-0b3c1798-c722-4688-a46f-b0697b1e0f95.jpg)

**Comment by hufman** [comment]
Thank you for testing, and I'm pleased you are enjoying it!

Automatic rerouting and turn-by-turn directions are hard, because I'm not actually sure how to locate the user along the route! I also worry that bugs in automatic rerouting may explode the number of route requests that are sent to the Mapbox service (Mapbox only gives 100,000 per month for free, which I will eventually need to spread among all of the app store users), and so the manual recalculation is a suitable compromise for this version. 

I believe Mapbox includes a small search history, but I count this as part of the saved places from #4.

Rotating the map, similar to tilting the map, has complications because of the low update framerate over Bluetooth. I worry about the user experience of the map jumping around and showing a different arbitrary angle as it updates at 1fps.
Adjusting the camera panning based on direction of travel is much more feasible! I've added it as #540.

I don't have control over the information displayed on the map, unfortunately, that is provided by Mapbox.
The Build Variant code layout means that it is impossible to build the app with both Gmap and Mapbox included. Besides, anyone who is able to build the Gmap version is unlikely to also want the inferior Mapbox information :) There's also tricky implications about sharing data: For example, the Gmap license disallows showing Google Place and Routing search results on any non-Google map.

Thank you for your help in testing!

**Comment by hufman** [comment]
@fata74 Thank you for testing! The map can only be displayed behind the other car UI elements (such as the status bar and the left navigation widgets). We might be able to expand it a little more to the left (@jezikk82 can you experiment and tell me the RHMIDimensions paddingLeft increase that is needed in your ID4 car?), but that is all that can be done.
I should hide the Show Widescreen option if the car screen isn't wide enough to support it, to reduce confusion.

**Comment by jezikk82** [comment]
In #509 I asked how to build artefacts. RHMI dimensions are there. I can experiment with that. You can try to decrease left padding for id4 cars and see.

**Comment by hufman** [comment]
I updated the dev docs to provide instructions which should help you test out RHMIDimensions changes, try it out! I don't have an ID4 car to test with, so I'll need to rely on your contributions.

**Comment by izzy-xd** [comment]
Hi. Maybe it will help you. Thanks for a good job :) 
![Screenshot_20220225-170551__01](https://user-images.githubusercontent.com/28729185/155751751-618656ba-0f2b-49f0-b18b-69e2b1b9023e.jpg)
![IMG_20220225_170508](https://user-images.githubusercontent.com/28729185/155751755-02bcd9f9-39c9-4ee4-b44a-013fc0beeb38.jpg)

**Comment by berseker** [comment]
Hi hufman, my contribution with my experience so far and some minor improvements\bugfixes needed:

I'm trying the mapbox navigation while commuting & my experience is that using at the same time
- Musicolet with MP3 streaming
- Mapbox
often causes a "sudden" death of the app: 
maybe the two above functionalities with only bluetooth bandwith are a bit too much.

music stops, the display freezes on the map, I can change BMW screen to some other functionality but bluetooth connectivity remain "halted", I have close the unit navigation and wait that it completely turns off in order to make it work again. I've tried to get the logs but i do not why the logcat I take lately have only a few minutes of history. I'll try again in the next days.

> hmi.display-height: 480
hmi.display-width: 800
hmi.role: HU
hmi.type: BMW ID6L
hmi.version: EntryEvo_ID5_1903_Release ID5_1903-490-1837K Build 47 - Rev:203015 2018-11-14 08:39:42

![screen_1](https://user-images.githubusercontent.com/1159431/156050006-03033ea2-d4ea-4cea-8496-6d5d3cb9bb96.jpg)
![screen_3](https://user-images.githubusercontent.com/1159431/156050009-f58f32e1-e3e9-4ae6-b7ea-438188b7554d.jpg)

**Comment by hufman** [comment]
Thank you for the testing and feedback, @berseker!
The crashing sounds like it is similar to #536. I unfortunately haven't experienced it, so I must try to read the donated logs and see if there's anything that can be improved.
Have you somehow modified your IDrive at all? I wonder why your car is saying it has a resolution of 800x480, which is (for example) the size of [fata74's screen](https://github.com/BimmerGestalt/AAIdrive/issues/538#issuecomment-1049829217). Your screen looks more like [izzy-xd's screen](https://github.com/BimmerGestalt/AAIdrive/issues/538#issuecomment-1051007203), which is 1280x480.
The Screen Mirroring and this Maps module use your car's claimed screen resolution to determine how big of an image to send (the car doesn't seem to support any native stretching operations). Because your car claims to have a narrow screen ([less than 900](https://github.com/BimmerGestalt/IDriveConnectKit/blob/8b47bca3c858b3b4cd8c36550afe2081c2599ca1/src/main/java/io/bimmergestalt/idriveconnectkit/RHMIDimensions.kt#L121) is the threshold) the widescreen button won't change anything. However, this screen resolution would not provide such a large image in the first picture, which is very confusing to me. I haven't seen any other instances of the car's claimed resolution being so incorrect!
Can you share more information about your IDrive version?

**Comment by berseker** [comment]
.... you may be right, originally I had a 6,5" screen, then I bought an 8,8" screen from an F30 and codified it myself. but any other window of ID6 works properly. AFAIK the 1280x480 resolution are for 10.25" inches of other models like BMW X3\X5 and so on

I just checked And I (should) have coded
DISPLAY_SIZE_RESOLUTION = 8.8_1280x480
DISPLAY_VARIANTE = breites_display

so I think something is wrong in my coding.. I will further investigate

**Comment by hufman** [comment]
Oh good, that's at least an explanation! I know my 2021 Mini has an 8.8" screen that is technically 1440x540 but it claims to only be 1280x480, so I have a [workaround](https://github.com/BimmerGestalt/IDriveConnectKit/blob/8b47bca3c858b3b4cd8c36550afe2081c2599ca1/src/main/java/io/bimmergestalt/idriveconnectkit/RHMIDimensions.kt#L36) just for Minis (perhaps because of their curved screens), and was worried I'd have to figure out more workarounds! The previous Mini ID4 XL screen was 8.8" and 1280x480.

**Comment by berseker** [comment]
... i've checked and I have to apologize.. that "screenshot" was taken When I had the old screen. the new one shows this
![image](https://user-images.githubusercontent.com/1159431/156426200-c9e80bfa-aa1b-4224-857d-b364158d1a90.png)
but that means that maybe on ID6 there are some tricks to do to "expand" the full screen to the right size...

**Comment by hufman** [comment]
@fata74 @izzy-xd Please try this [test build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-dimensions-test-mapbox-sentry-release.apk) which changes the left padding and so the image should fill up the left side better.

@berseker Do you know the actual pixel resolution of the screen you installed? It seems that [even IDrive 5](https://www.bimmerfest.com/threads/all-new-bmw-idrive-5-technical-details.893672/) had the possibility of running at 1440x540, so I wonder if you might try coding your car to be a 8.8" 1440x540 resolution. What coding options do you see?

**Comment by berseker** [comment]
@hufman i'll check about It.. actually the part Number of my 8.8" display Is 6822626, thats the only thing i know up until now

**Comment by hufman** [comment]
I looked around for that part number and the forum posts all say 1280x480, I didn't find a definitive answer about whether it was potentially actually a higher res...

Another [test build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-dimensions-sidebar-mapbox-sentry-release.apk) removes the ID4 specific changes from the previous test build, because I think I added an extra change that may have corrected the padding problem without adding ID4-specific values, so I'd appreciate testing with that version too!

**Comment by izzy-xd** [comment]
![IMG_20220308_081001](https://user-images.githubusercontent.com/28729185/157209318-7d57b794-73b7-4544-b136-c71c155cd2e3.jpg)
![Screenshot_20220307-200939__01](https://user-images.githubusercontent.com/28729185/157209338-3718bdde-4014-4916-8982-dd6bb5f14500.jpg)

**Comment by fata74** [comment]
Hi, I have tested it but it looks like before to me.
![PSX_20220308_103003](https://user-images.githubusercontent.com/56363659/157210400-31244efe-0de6-40e9-8083-9db5e852e240.jpg)

**Comment by fata74** [comment]
I have tried this one:
androidautoidrive-latest-dimensions-test-mapbox-sentry-release.apk

**Comment by izzy-xd** [comment]
![Screenshot_20220308-103738__01](https://user-images.githubusercontent.com/28729185/157276910-5210a5ab-7255-4279-af65-72e0d02f2ad7.jpg)
![IMG_20220308_164554](https://user-images.githubusercontent.com/28729185/157276941-cc6771b8-c330-4b2e-a036-79412ea4aea2.jpg)

**Comment by magamess** [comment]
Hello @hufman, first at all, thank you very much for all your effort on this project, it is amazing what you have achieved.

I know there is separate discussion on #517, but I wonder if it possible to get the same layout (size) @izzy-xd is showing now for Mapbox but for the screen mirroring app. I am using screen mirroring with Android Auto and I am still getting big black margins. Is there anything that can be adjusted to reduce the margins? 

My NBT resolution is 1280x480
My mobile resolution is 2400 x 1080

Thank you very much!

**Comment by hufman** [comment]
@izzy-xd did you uncheck the Widescreen Map setting? I haven't found any way to determine if the car is running in widescreen mode or sidebar mode, so this has to be done manually. I want to know if the sidebar mode lines up nicely with the edge of the screen when your car is showing a sidebar, or if it wastefully draws under the sidebar. I'm especially curious if the dimensions-testing version looks right in this regard (lined up perfectly in my Mini), because I like how the dimensions-testing version fills the screen more completely but worry that it would fill too much under the sidebar.

@magamess As mentioned in the main discussion thread in #517, because AAIdrive is in control of its map, it can adjust the picture without the car's overlays drawing over any important parts of the map. The same negative padding could be applied for mirroring, but might cause image obstructions and so I don't feel comfortable applying it to the generic mirroring addon. But please, let's continue this discussion in the screen mirroring discussion section.

**Comment by izzy-xd** [comment]
![Uploading IMG_20220310_082640__01.jpg…]()

**Comment by hufman** [comment]
Thank you for your testing and feedback! I've pushed up these changes to the main build.

**Comment by HaivanJV** [comment]
> Rotating the map, similar to tilting the map, has complications because of the low update framerate over Bluetooth. I worry about the user experience of the map jumping around and showing a different arbitrary angle as it updates at 1fps.

Hi Hufman. First of all, thanks as usual for all the hard work and the amazing results.
I have a point to make about the above. Would it be possible to still try implementing it on a test build?
Personally, I am not able to follow any navigation system that doesn't offer this feature. I am getting lost when navigation is not pointing at driving direction.

Contextually, I tested Gmaps in screensharing via Android Auto app: results under bluetooth connection were acceptable and navigating inside the city was usable. Connected via USB was nearly perfect.

I think I have read on a different issue that, under Android Auto, the screensharing of Gmaps has a better FPS. Isn't it the same or a similar rate of the Mapbox integration?

**Comment by hufman** [comment]
I pushed a [test build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-bmw_xl-mapbox-sentry-release.apk) where any EntryEvo_ID5 car claiming to be 1280px wide will actually run as if it has a 1440x540 screen. Try that out @berseker

**Comment by berseker** [comment]
@hufman "We have a Winner!"
![PXL_20220401_073548415_1](https://user-images.githubusercontent.com/1159431/161203036-efb57d34-fdd8-46e5-92f4-1ad8d8bf049a.jpg)

**Comment by hufman** [comment]
Ah ha, the image is too wide and the compass was cut off! I pushed a [new version](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-bmw_xl-mapbox-sentry-release.apk) that should adjust the border padding to line up perfectly.

@unlive74 can you post a picture of your EntryEvo_ID5 showing the map (preferably the main build, not the bmw_xl test build), and tell me what hmi.display-width the car reports? I'm trying to determine the rule for when to force the image to be bigger!

**Comment by hufman** [comment]
@jezikk82 can you check the map size in your car (with pictures so I can measure and adjust if needed), I believe you also have a bigger screen size and I wonder if any of the screen measurements need adjusting.

**Comment by berseker** [comment]
> Ah ha, the image is too wide and the compass was cut off! I pushed a [new version](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-bmw_xl-mapbox-sentry-release.apk) that should adjust the border padding to line up perfectly.
> 
> @unlive74 can you post a picture of your EntryEvo_ID5 showing the map (preferably the main build, not the bmw_xl test build), and tell me what hmi.display-width the car reports? I'm trying to determine the rule for when to force the image to be bigger!

maybe a fine tuning it's necessary, but the compass actually appeared when I started moving with the car (if i remember correctly) in any case i'll try also this new pushed release ;)

**Comment by jezikk82** [comment]
> @jezikk82 can you check the map size in your car (with pictures so I can measure and adjust if needed), I believe you also have a bigger screen size and I wonder if any of the screen measurements need adjusting.

Sorry for late reply, this is how it looks (latest version, my own built) in my car 
![AAID_1440](https://user-images.githubusercontent.com/7836767/161919685-f9ef1226-8dc2-4172-b925-e6e50b5dc12c.jpg)

AAID report resolution of 1440x540, NTBevo_ID5_1911_*, a4axl:false. NBT pro, biggest screen. According to BMW spec it should be 1920px wide screen, but according to my measure is about 1600px wide.

**Comment by jezikk82** [comment]
Here is the variant with my adjustments I used with GMaps:
![AAID_1440_myAdj](https://user-images.githubusercontent.com/7836767/161929209-4c366515-9cf6-4b4c-9c0a-e7a7b89f3368.jpg)

Values are:
val rhmiWidth: Int = 1550  
val rhmiHeight: Int = 550  

val marginLeft: Int = 110  
val marginRight: Int = 5

val paddingLeft: Int = 190 
val paddingTop: Int = 70   

It's still require small corrections but I don't mind to have it as in pic.  

p.s. During "navi" I dont see any arrow showing directions or next turn - but I saw in commits history @hufman added it ?
Any change to add auto recalc option after x seconds/meters of detour  ?

**Comment by unlive74** [comment]
hmi.display-height: 480
hmi.display-width: 1280

![20220406_182640](https://user-images.githubusercontent.com/75434533/161990019-bd36d71a-13e3-400b-8deb-611edf9b232c.jpg)
![20220406_182731](https://user-images.githubusercontent.com/75434533/161990056-59273f24-beeb-45cd-80ce-fecf47dbd38d.jpg)

Btw: Build 320c3a4 won't connect to my car. build 9dfe420 works great.

**Comment by unlive74** [comment]
@hufman OK I'll try tomorrow. unfortunately for me it's just a test - the build with the map is irrelevant for me due to frequent crashes. At the same time, it also hangs the system (only rebooting EntryEvo helps).

**Comment by berseker** [comment]
@hufman here is my test . I tried also mirroring plugin but I think you have made only the mod in the mapbox plugin for now
![image](https://user-images.githubusercontent.com/1159431/162277957-2a68796c-d209-4e5d-8ced-3300d8f2032f.png)

**Comment by hufman** [comment]
Wonderful, thank you so much! You are correct, Screen Mirroring hasn't been changed yet, and I see from the code that it is using the Generic dimensions which move the screen up 80 pixels, while the dimensions specific to your screen would only move it up 60 pixels and should fix that.
The compass is tricky, I've tried applying a margin to it and it doesn't seem to move. It's even worse in the Mini because of the curved screen :)

**Comment by berseker** [comment]
In any case please note that We are Talking about details to get to perfection, which Is really really close by now!

**Comment by jezikk82** [comment]
@hufman  Is there a way to overwrite object RHMIDimensions from class SidebarRHMIDimensions using the same way you posted in test build (in MapServiceApp.kt). This would allow the AAID to be built without compiling the IDriveConnectKit every time I play with something. Thanks.

**Comment by hufman** [comment]
Of course! You can build whatever custom RHMIDimensions subclass you'd like, [following this example](https://github.com/BimmerGestalt/AAIdrive/commit/52ea6bd0), which doesn't require switching IDriveConnectKit from a gradle artifact to being a submodule. If you mean you want to tweak the SidebarRHMIDimensions, you can edit MapAppMode to return a custom subclass instead of the library's SidebarRHMIDimensions easy enough.
Would that work for you, or what else are you looking to do?

**Comment by hufman** [comment]
I posted a [debug branch](https://github.com/BimmerGestalt/AAIdrive/tree/custom_dimensions) and [build](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-custom_dimensions-mapbox-nonalytics-release.apk) that provides textboxes in the Settings tab to override the RHMIDimensions, which takes effect by closing and opening the map again instead of needing to recompile and reconnect to the car. @jezikk82 maybe that will help speed up your testing!

**Comment by jezikk82** [comment]
Thanks, for sure it will help. I'll have access to 2 other bmws from f and g series with different i drives so I will try to adjust the mapbox as close as possible. Let me know if  you prefer map on the left flot to border of the screen or end up before home button (delicate grey line in background noticeable).

**Comment by jezikk82** [comment]
I tested with 2 different BMW G series with NBTevo ID5 1440x

margin left: 0
margin right: 5
padding left: 90
padding top: 67
image size must be reduced to: 1345 x 540 
![IMG_20220412_203854](https://user-images.githubusercontent.com/7836767/163034423-3d0144d5-890d-43eb-84be-010896f79970.jpg)

When widescreen  is off map will look like this:
![IMG_20220412_203830](https://user-images.githubusercontent.com/7836767/163034570-d41a4082-7fa3-4451-a1b0-29625a423584.jpg)
![IMG_20220412_203803](https://user-images.githubusercontent.com/7836767/163034575-4bb02c84-fcc6-4791-9708-ee2b4626dfd9.jpg)

As you can see sidebar is covering around 20px of the map. 

Another variant is as follow:
![IMG_20220412_155847](https://user-images.githubusercontent.com/7836767/163034703-66392107-b2b0-458b-ac6c-d20a6f729d77.jpg)
![IMG_20220412_154423](https://user-images.githubusercontent.com/7836767/163034765-ce8d20dd-3007-4b5c-b019-36c5ea163c46.jpg)

width:1350
height: 480
margin left: 0
margin right: 5
padding left 90
padding top 0

Also heve sidebar is covering about 20px of the map. 

Later this week I'll have change to test it on f&g series with smaller navi (entryNav). Will post the results. 

Let me know if you need anything else to test here ?

**Comment by hufman** [comment]
Cool, thank you so much for the precise measurements! I updated the custom_dimensions build with the numbers, and I added a -20px offset for not-widescreen mode of that generation of screen.
The width adjustment is done by increasing the marginLeft, which represents the left area on the screen that isn't used by the map, and the marginLeft and marginRight together reduce rhmiWidth into the visibleWidth that is used by the map.
